### PR TITLE
[All models] Extend config.output_attentions with output_attentions function arguments

### LINKED ISSUE #3880: Replace `config.output_attentions` parameter with function argument `output_attentions`
# 🚀 Feature request

Currently the user has to decide whether the model should output the attentions when she/he creates the config of a model: config.output_attentions = True/False. It would be nice if the user can decide this when calling the models `forward()` / `call()` with a flag `output_attentions`. This should be done for all TF and PT models that can output attentions.

A very similar recent change was done for the variable `config.output_past` -> see PR:#3734

## Motivation

The user has more flexibility when the hidden states should be output or not.

## Your contribution

If someone feels like contributing to the library, this would be a great first PR. I'm very happy to guide the contributor through the PR!

### PR DESCRIPTION
**Pull request by bharatr21**
Attempts to close #3880 by refactoring `config.output_attentions` => `output_attentions` in `forward()`/`call()` functions

**UPDATE** from Patrick

@thomwolf @LysandreJik 
This PR adds the argument `output_attentions` to every forward function for more flexibility. This PR makes it possible to easily switch the output attention on/off without having to instantiate a new model every time.

The logic is the following. If 'output_attentions` is configured in the `forwrd()` fn => use that. If not => use `config.output_attentions`.

IMPORTANT: This PR does **not** change backward compatibility since output_attentions can still be configured using the config.

TESTS:
An additional test is added to the `test_output_attentions()` common test.

FUTURE PR:
- [ ] Clean the documentation. We still need to add this argument to the docs of all models and make sure the docs are clean. Lemme know @Bharat123rox if you want to tackle this in a new PR or if I should do it :-) It's not the most interesting PR, so I fully understand if you don't feel like doing it anymore ;-)

### CODE REVIEW

**Comment by Bharat123rox** [comment]
Also removed the tests for `output_attentions` since all of them were fetching values from the config

File: src/transformers/modeling_ctrl.py
Comment (patrickvonplaten) [inline src/transformers/modeling_ctrl.py:278]: `output_attentions` should not be saved in any model class - we should be able to always switch `output_attentions` on and off in the forward pass

File: src/transformers/modeling_transfo_xl.py
Comment (patrickvonplaten) [inline src/transformers/modeling_transfo_xl.py:555]: Also let's not save `self.output_attentions` here

File: tests/test_modeling_common.py
Comment (patrickvonplaten) [inline tests/test_modeling_common.py:422]: In this test we should make sure that in the  forward function `output_attentions=True` . We can just add `inputs_dict["output_attentions"] = True` to the input.

File: tests/test_modeling_common.py
Comment (patrickvonplaten) [inline tests/test_modeling_common.py:391]: Same as below add `output_attentions=True` to the `inputs_dict`

File: tests/test_modeling_common.py
Comment (patrickvonplaten) [inline tests/test_modeling_common.py:204]: I'm ok with this!

File: tests/test_modeling_common.py
Comment (patrickvonplaten) [inline tests/test_modeling_common.py:133]: Basically, in every test which sets `config.output_attentions = True` we should instead add this to the forward function to have the same result.

File: tests/test_modeling_tf_common.py
Comment (patrickvonplaten) [inline tests/test_modeling_tf_common.py:299]: Let's not care about TF in this PR and do this in another one.

Comment (patrickvonplaten) [inline tests/test_modeling_tf_common.py:299]: The same thing applies here as in the pytorch tests.

**Comment by patrickvonplaten** [comment]
@Bharat123rox - that's awesome, thanks a lot! It's a lot of manual work, but it will be super useful once it's merged :-)

I added a bunch of comments - let me know if something is unclear.

Regarding the tests, let's try to first make all torch tests pass and then check the TF tests. 

Regarding the torch tests:

Most tests that fail are those where you removed `config.output_attentions` in the test, but didn't set `output_attentions=True` for the forward call. These tests previously were outputting the attentions but don't do this anymore. You should fix these tests if you set `output_attentions=True` in the forward pass. 

Looking forward to have this merged soon :-) Let me know if something is unclear!

**Comment by Bharat123rox** [comment]
Now, most of the tests are giving a new error:
`> E TypeError: forward() got an unexpected keyword argument 'output_attentions'`

**Comment by patrickvonplaten** [comment]
OK! Let's maybe try first to fix all the test `test_attention_outputs` tests. You can run this test for a specific model using the following command: 
```
pytest tests/test_modeling_openai.py::OpenAIGPTModelTest::test_attention_outputs
```
I fixed this test as an example for `openai` on this commit: https://github.com/huggingface/transformers/pull/4597/commits/e8efd72fce1be304043863cbab4cd7a61a39e434

It's gonna be a longer process to fix all tests. Let's try to start with the `test_attention_outputs` tests for all PyTorch models :-) 

Btw, can you do the following change to the branch `outputattentions` of your fork, so that I can commit directly to your branch? :-) 

https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork

**Comment by Bharat123rox** [comment]
https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork was not clear as I've already checked the "Allow edits by maintainers" option, however I have sent a collaboration invite to you @patrickvonplaten which I think should give enough permissions

**Comment by Bharat123rox** [comment]
@patrickvonplaten the remaining failures in `test_output_attentions` of Torch tests are left because I tried to introduce `output_attentions` in `BertEncoder` which led to further downstream failures in torch modules (such as `AttributeError: 'bool' has no attribute 'dim'`)

Please guide me on what to do next, and how to fix failures in TF tests as well

**Comment by Bharat123rox** [comment]
@patrickvonplaten @thomwolf please help me in fixing the remaining PyTorch and TensorFlow test failures, they are of various kinds, mostly `AssertionErrors`

File: src/transformers/modeling_bart.py
Comment (patrickvonplaten) [inline src/transformers/modeling_bart.py:370]: any specific reason to call this argument `need_weights`? Could we maybe change it also to `output_attentions` for consistency?

File: src/transformers/modeling_bart.py
Comment (patrickvonplaten) [inline src/transformers/modeling_bart.py:212]: also please rename to `output_attentions`

**Comment by patrickvonplaten** [comment]
This is really great work! I will take a look at the failing tests :-) It would be great if you can rename some of the variables as mentioned above.

File: src/transformers/modeling_bart.py
Comment (Bharat123rox) [reply src/transformers/modeling_bart.py:372]: It was called `need_weights` earlier also and the only reason I didn't touch it was that I was not aware of any downstream impact 😅 , I'll be fixing it now

File: src/transformers/modeling_roberta.py
Comment (patrickvonplaten) [inline src/transformers/modeling_roberta.py]: I think this was deleted accidently

File: src/transformers/modeling_longformer.py
Comment (patrickvonplaten) [inline src/transformers/modeling_longformer.py]: I think `masked_lm_labels` was deleted accidently here

File: src/transformers/modeling_electra.py
Comment (patrickvonplaten) [inline src/transformers/modeling_electra.py]: I think `masked_lm_labels` was deleted accidently here

File: src/transformers/modeling_distilbert.py
Comment (patrickvonplaten) [inline src/transformers/modeling_distilbert.py]: I think `masked_lm_labels` was deleted accidently here

File: src/transformers/modeling_albert.py
Comment (patrickvonplaten) [inline src/transformers/modeling_albert.py]: I think `masked_lm_labels` was deleted accidently here

File: src/transformers/modeling_roberta.py
Comment (Bharat123rox) [reply src/transformers/modeling_roberta.py]: No, it was merely renamed to `labels`, so this change seems to be correct. Unfortunately this is showing up separately after I fixed conflicts in the merge commit

File: src/transformers/modeling_bart.py
Comment (patrickvonplaten) [reply src/transformers/modeling_bart.py]: True! Thanks for renaming it :-) We should have a consistent name across models

File: src/transformers/modeling_longformer.py
Comment (Bharat123rox) [reply src/transformers/modeling_longformer.py]: No it was renamed to `labels`

File: src/transformers/modeling_electra.py
Comment (Bharat123rox) [reply src/transformers/modeling_electra.py]: Renamed to `labels`

File: src/transformers/modeling_roberta.py
Comment (patrickvonplaten) [reply src/transformers/modeling_roberta.py]: gottcha! Never mind my comment then :-)

File: src/transformers/modeling_distilbert.py
Comment (Bharat123rox) [reply src/transformers/modeling_distilbert.py]: Renamed to `labels`

File: src/transformers/modeling_albert.py
Comment (Bharat123rox) [reply src/transformers/modeling_albert.py]: Renamed to `labels`

**Comment by patrickvonplaten** [comment]
Ok great, I can take a look into the remaining tests now :-)

**Comment by Bharat123rox** [comment]
> Ok great, I can take a look into the remaining tests now :-)

Yes, please do, thank you! there are only 3 Assertion failures in Torch and hopefully all failures in TF are also similar 🤞

File: src/transformers/modeling_tf_bert.py
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_bert.py:26]: @Bharat123rox here you can see how `output_attentions` should be applied to a TF model.

You have to do the following:
1) Make sure that `output_attentions` is passed to all forward functions as a list parameter
2) cast all `output_attentions` booleans using the function `cast_bool_to_primitive`
3) extract `output_attentions` from a dict or list if necessary

File: src/transformers/modeling_tf_bert.py
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_bert.py:275]: Here 2)  output_attenitons is casted to prim bool

File: src/transformers/modeling_tf_bert.py
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_bert.py:312]: Here 1) output_attentions is passed as a list argument

File: src/transformers/modeling_tf_gpt2.py
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_gpt2.py:684]: Here 3) output_attentions is extracted from list

File: src/transformers/modeling_tf_gpt2.py
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_gpt2.py:696]: Here 3) output_attentions is extracted from dict

**Comment by patrickvonplaten** [comment]
Hey @Bharat123rox,

I think from the PyTorch side we are ready now :-) 
Regarding the TF side, my last commit shows how to implement it for TF Models - could you give it a try for the other models? You always have to pay attention to points 1) - 3) as mentioned above to make sure that the TF Models can be trained, complised and serialized with keras.

Be careful to make a `git fetch` and `git pull` on your branch now before continuing to work since I added two commits. 

Let me know if you have any questions! :-) 
Really great work so far, I think we are almost finished :-)

**Comment by Bharat123rox** [comment]
Remaining TF test failures mostly fail here in `test_tf_common.py`:
```
>           self.assertEqual(len(attentions), self.model_tester.num_hidden_layers)
E           AssertionError: 13 != 5
```

**Comment by Bharat123rox** [comment]
@patrickvonplaten Most of the TF tests are fixed, and the remaining seem to be different `AssertionErrors`, please help with the remaining TF test failures

File: src/transformers/configuration_utils.py
Comment (patrickvonplaten) [inline src/transformers/configuration_utils.py:58]: @thomwolf @LysandreJik @sshleifer - `output_attentions` is completely removed as an argument to the config. This means that `output_attentions` always defaults to `False` in the forward functions and cannot be influenced anymore by the `config`. 

It is arguable whether we should instead allow to initialize `config.output_attentions` with a boolean and let all `output_attentions` in all `forward()` functions default to `None` and instead take the value from `config.output_attentions`, but there would be some problems here as well:

1) It adds a lot of code
2) It's probably confusing to the user to read in the docs that the boolean `output_attentions` argument is init to `None` for all forward functions, but behaves as init to `False`. 
3) At which abstraction level do we still allow the `output_attentions` to be initialized from the config? `BertModel.forward()` should probably be initialized with `config.output_attentions`, but `BertAttentionLayer()` as well? How long do we have to pass down the config?
4) This could also introduce the following confusion: Contrary to the `generate()` method which directly relies on the `self.config` config, some models, e.g. T5 make a copy of the original config and then pass the config further down. This could lead to confusion when people initially instantiate a T5 model with `config.output_attentions=True` and try to change it afterward with `model.config.output_attentions=False`, which would not work, since the model classes are already initiated with `config.output_attentions=True`. 

All in all, I would be in favor of removing `config.output_attentions` and initialize `output_attentions` only via the argument `output_attentions=False` in all forward functions. I think the same should apply for `use_cache` and `output_hidden_states` (`use_cache` is already completely independent for all `forward()` methods and only the `generate()` method relies on the config, which I think we could change as well. 
IMO `use_cache`, `output_attentions` and `output_hidden_states` all are not really parameters of the model or its decoding strategy, but switches for all (or some for `use_cache`)models that can be turned on and off, but essentially don't influence the model's output logits. In contrast the other config attributes ("max_length", "do_sample", ...) do influence the output of the model and therefore should stay in the config. Therefore I would be ok with removing `use_cache`, `output_attentions` and `output_hidden_states` completely from the `config`. 

But that's up for debate!

What is your opinion on that?

File: src/transformers/modeling_tf_bert.py
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_bert.py:279]: @thomwolf @LysandreJik we need to add a `cast_bool_to_primitive` function in all tf files to deal with keras serialization and compilation

File: src/transformers/modeling_tf_utils.py
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_utils.py:1750]: Needed for keras serialization and tf compilation

File: src/transformers/modeling_tf_xlnet.py
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_xlnet.py:1445]: This dead code here could be removed in another PR

File: tests/test_modeling_common.py
Comment (patrickvonplaten) [inline tests/test_modeling_common.py:204]: Although it never looks good to delete a test :D, I think it's fine here:

- `self._create_and_check_torchscript` is already tested with `input_dict["output_attentions"] = True` in the next test, so this test does not create any additional test coverage in my opinion.

**Comment by sshleifer** [comment]
I got an email with a comment from @patrickvonplaten  (which I can't find) expressing the opinion that 
>IMO use_cache, output_attentions and output_hidden_states all are not really parameters of the model or its decoding strategy, but switches for all (or some for use_cache)models that can be turned on and off, but essentially don't influence the model's output logits. In contrast the other config attributes ("max_length", "do_sample", ...) do influence the output of the model and therefore should stay in the config. Therefore I would be ok with removing use_cache, output_attentions and output_hidden_states completely from the config.

And I completely agree with that conclusion! We should make sure to highlight it in release notes.

**Comment by patrickvonplaten** [comment]
> I got an email with a comment from @patrickvonplaten (which I can't find) expressing the opinion that
> 
> > IMO use_cache, output_attentions and output_hidden_states all are not really parameters of the model or its decoding strategy, but switches for all (or some for use_cache)models that can be turned on and off, but essentially don't influence the model's output logits. In contrast the other config attributes ("max_length", "do_sample", ...) do influence the output of the model and therefore should stay in the config. Therefore I would be ok with removing use_cache, output_attentions and output_hidden_states completely from the config.
> 
> And I completely agree with that conclusion! We should make sure to highlight it in release notes.

@sshleifer 
I just deleted this comment :D I rethought this a bit. I think the better solution is what we have now: Have `output_attentions` in both the config and as a forward argument. This way we can still use the keras serialize function, don't break backward compatibility and have the same logic as we do in the generate() method.

**Comment by patrickvonplaten** [comment]
@Bharat123rox - thanks a million for your work here! It was a lot of manual work in a lot of files! This PR is VERY useful for the library!

File: src/transformers/modeling_albert.py
Comment (patrickvonplaten) [inline src/transformers/modeling_albert.py:535]: @thomwolf  @LysandreJik @sshleifer - This is written in every forward function so that we have the logic as explained in the description

**Review by patrickvonplaten — APPROVED** [review]

File: src/transformers/modeling_flaubert.py
Comment (sshleifer) [inline src/transformers/modeling_flaubert.py:293]: is this still on `config`? Is that desired behavior?

**Review by sshleifer — APPROVED** [review]
Epic! Left one comment about `config.hidden_states`, which IMO should be treated the same way. Can be a separate PR thought :)

File: src/transformers/modeling_flaubert.py
Comment (patrickvonplaten) [reply src/transformers/modeling_flaubert.py:293]: yeah the 'output_hidden_states` is going to be replaced in a separate PR: https://github.com/huggingface/transformers/pull/4548 .

It's prob going to be quite a mess with merge conflicts so we will merge this one first and then I rebase the other one

File: src/transformers/configuration_utils.py
Comment (LysandreJik) [inline src/transformers/configuration_utils.py:58]: If  we remove the `config.output_attentions` from the documentation, then I think we should also add a deprecation warning so that users are aware this isn't the desired way to control the attentions anymore.

File: tests/test_modeling_common.py
Comment (LysandreJik) [inline tests/test_modeling_common.py:152]: Great to still test this!

**Review by LysandreJik — APPROVED** [review]
This is great. Thanks for the hard work @Bharat123rox!

**Review by thomwolf — APPROVED** [review]
Ok, huge work @Bharat123rox!
LGTM!

**Pull request merged**
